cmd/deploy.go

```diff
@@ -131,44 +161,85 @@ func createReleaseFile() (filename string, err error) {
-		// git ls-files -d // delete files
-		output, err = exec.Command("git", "ls-files", "-d").CombinedOutput()
-		if err != nil {
-			panic(fmt.Errorf("Failed listing deleted "))
-		}
-		tarball.SetIgnoreList(strings.Split(string(output), "\n"))
+	if _, errGit := os.Stat(".git"); !hasGit || os.IsNotExist(errGit) {
+		// not a GIT repo/environment!
+		d.Println("Fallback to tarball full current working directory...")
+		cwd, _ = os.Getwd()
+		filename, err = tarball.CompressFolder(cwd)
+		return
+	}
+
+	// we are in a GIT environment!
+	var (
+		files, allFiles []string
 
-		// Include list
-		// git ls-files -c
-		output, err = exec.Command("git", "ls-files", "-c").CombinedOutput()
-		if err != nil {
-			panic(fmt.Errorf("Failed list Git cached files"))
+		gitListingFilesFlags = [][]string{
+			// Include commited files - git ls-files -c
+			{"-c"},
+
+			// Untracked files - git ls-files -o --exclude-standard
+			{"-o", "--exclude-standard"},
 		}
-		files = append(files, strings.Split(string(output), "\n")...)
-		// git ls-files -o --exclude-standard
-		output, err = exec.Command("git", "ls-files", "-o", "--exclude-standard").CombinedOutput()
-		if err != nil {
-			panic(fmt.Errorf("Failed list Git untracked non-ignored files"))
+	)
+
+	// Exclude list - git ls-files -d
+	if files, err = d.parseFilesListFromGIT([]string{"-d"}); err != nil {
+		return
+	}
+	tarball.SetIgnoreList(files)
+
+	for _, lsArgs := range gitListingFilesFlags {
+		if files, err = d.parseFilesListFromGIT(lsArgs); err != nil {
+			return
 		}
-		files = append(files, strings.Split(string(output), "\n")...)
 
-		filename, err = tarball.CompressFiles(files)
-	} else {
-		fmt.Println("Fallback to tarball full current working directory...")
-		cwd, _ = os.Getwd()
-		filename, err = tarball.CompressFolder(cwd)
+		allFiles = append(allFiles, files...)
 	}
+	filename, err = tarball.CompressFiles(d.handleDeployEnv(allFiles))
+	return
+}
 
+func (d *KoolDeploy) parseFilesListFromGIT(args []string) (files []string, err error) {
+	var output string
+
+	output, err = d.Exec(d.git, append([]string{"ls-files", "-z"}, args...)...)
+	if err != nil {
+		d.Error(err)
+		err = fmt.Errorf("failed listing GIT files")
+		return
+	}
+
+	// -z parameter returns the utf-8 file names separated by 0 bytes
+	files = strings.Split(output, string(rune(0x00)))
 	return
 }
+
+// handleDeployEnv tackles a special case on kool.deploy.env file.
+// This file can or cannot be versioned (good practice not to, since
+// it may include sensitive data). In the case of it being ignored
+// from GIT, we still are required to send it - it is required for
+// the Deploy API.
+func (d *KoolDeploy) handleDeployEnv(files []string) []string {
+	path := filepath.Join(d.envStorage.Get("PWD"), koolDeployEnv)
+	if _, envErr := os.Stat(path); os.IsNotExist(envErr) {
+		return files
+	}
+
+	var isAlreadyIncluded bool = false
+	for _, file := range files {
+		if file == koolDeployEnv {
+			isAlreadyIncluded = true
```

Comment:
I guess if we return the files here, this variable will not be necessary and neither the if statement below

Reply:
I see, that is strictly correct. However, in this instance I rather keep the control variable... otherwise we mix-up the returning of `files` into an weird-loop structure that kind hides its original intent (simply check if the file already existed).